src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -19,80 +23,212 @@
+            verrors.add(
+                f'{schema_name}.certificate',
+                'Not a valid certificate'
+            )
         else:
-            cert_certificate = certificate_list[0]['cert_certificate']
+            cert_info = await middleware.call('certificate.load_certificate', certificate)
+            if not cert_info:
+                verrors.add(
+                    f'{schema_name}.certificate',
+                    'Certificate not in PEM format'
+                )
+
+    private_key = data.get('privatekey')
+    passphrase = data.get('passphrase')
+    passphrase2 = data.get('passphrase2')
+    if private_key:
+        if not (await middleware.call('certificate.load_private_key', private_key, passphrase)):
+            verrors.add(
+                f'{schema_name}.privatekey',
+                'Please provide a valid private key with matching passphrase ( if any )'
+            )
+        elif passphrase != passphrase2:
+            for f in ['passphrase', 'passphrase2']:
+                verrors.add(
+                    f'{schema_name}.{f}',
+                    'Pass phrase confirmation does not match'
+                )
+    elif passphrase or passphrase2:
+        verrors.add(
+            f'{schema_name}.passphrase',
+            'No private key specified for pass phrase'
+        )
+
+    if (
+        (certificate and private_key) and
+        all(k not in verrors for k in (f'{schema_name}.certificate', f'{schema_name}.privatekey'))
+    ):
+        public_key_obj = crypto.load_certificate(crypto.FILETYPE_PEM, certificate)
+        private_key_obj = await middleware.call('certificate.load_private_key', private_key, passphrase)
 
-        # getting fingerprint of certificate
         try:
-            certificate = crypto.load_certificate(
-                crypto.FILETYPE_PEM,
-                cert_certificate
+            context = SSL.Context(SSL.TLSv1_2_METHOD)
+            context.use_certificate(public_key_obj)
+            context.use_privatekey(private_key_obj)
+            context.check_privatekey()
+        except SSL.Error as e:
+            verrors.add(
+                f'{schema_name}.privatekey',
+                f'Private key does not match certificate: {e}'
             )
-        except Exception:
-            return None
-        else:
-            return certificate.digest('sha1').decode()
+
+    key_length = data.get('key_length')
+    if key_length:
+        if key_length not in [1024, 2048, 4096]:
+            verrors.add(
+                f'{schema_name}.key_length',
+                'Key length must be a valid value ( 1024, 2048, 4096 )'
+            )
+
+    signedby = data.get('signedby')
+    if signedby:
+        valid_signing_cas = await middleware.call(
+            'certificateauthority.query',
+            [
+                ('certificate', '!=', None),
+                ('privatekey', '!=', None),
+                ('certificate', '!=', ''),
+                ('privatekey', '!=', ''),
```

Comment:
Could you add another ('id', '=', signedby) here instead of iterating though them all again below? Or did I miss anything?

Reply:
Nopes, you didn't. I'll have this updated